src/compiler/compiler.rs

```diff
@@ -275,124 +296,205 @@ impl Compiler {
+                  storage: Arc<Storage>,
+                  key: String,
+                  miss_type: MissType)
+                  -> Box<Future<Item=(CompileResult, process::Output),
+                                Error=io::Error>>
+        where T: CommandCreatorSync,
+    {
+        let process::Output { stdout, .. } = preprocessor_result;
+        let start = Instant::now();
+
+        let compile = self.kind.compile(creator, self, stdout, &parsed_args, cwd, &pool);
+        Box::new(compile.and_then(move |(cacheable, compiler_result)| {
+            let duration = start.elapsed();
+            if compiler_result.status.success() {
+                if cacheable == Cacheable::Yes {
+                    debug!("[{}]: Compiled, storing in cache",
+                           parsed_args.output_file());
+                    // fall through
                 } else {
-                    debug!("[{}]: Compiled but failed, not storing in cache", out_file);
-                    Ok((CompileResult::CompileFailed, compiler_result))
+                    // Not cacheable
+                    debug!("[{}]: Compiled but not cacheable",
+                           parsed_args.output_file());
+                    return future::ok((CompileResult::NotCacheable, compiler_result)).boxed()
+                        as Box<Future<Item=_, Error=_>>
                 }
+            } else {
+                debug!("[{}]: Compiled but failed, not storing in cache",
+                       parsed_args.output_file());
+                return future::ok((CompileResult::CompileFailed, compiler_result)).boxed()
             }
-        }
+            let mut entry = match storage.start_put(&key) {
+                Ok(entry) => entry,
+                Err(e) => return future::err(e).boxed()
+            };
+            let write = pool.spawn_fn(move || -> io::Result<_> {
+                for (key, path) in &outputs {
+                    let mut f = try!(File::open(&path));
+                    try!(entry.put_object(key, &mut f).map_err(|e| {
+                        let msg = format!("failed to put object `{:?}` in \
+                                           storage: {}", path, e);
+                        io::Error::new(io::ErrorKind::Other, msg)
+                    }));
+                }
+                Ok(entry)
+            });
+            Box::new(write.and_then(move |mut entry| {
+                if !compiler_result.stdout.is_empty() {
+                    let mut stdout = &compiler_result.stdout[..];
+                    try!(entry.put_object("stdout", &mut stdout));
+                }
+                if !compiler_result.stderr.is_empty() {
+                    let mut stderr = &compiler_result.stderr[..];
+                    try!(entry.put_object("stderr", &mut stderr));
+                }
+
+                // Try to finish storing the newly-written cache
+                // entry. We'll get the result back elsewhere.
```

Comment:
This comment should probably be clearer that we don't want to block returning the compile result to the client while we write to the cache.